x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/privilege/ManageApiKeyConditionalClusterPrivilege.java

```diff
@@ -0,0 +1,156 @@
+        }
+        if (this.realms.contains("_self") && this.users.contains("_self")) {
+            if (this.realms.size() > 1 || this.users.size() > 1) {
+                throw new IllegalArgumentException(
+                        "both realms and users must contain only `_self` when restricting access of API keys to the owner");
+            }
+        }
+
+        this.requestPredicate = (request, authentication) -> {
+            if (request instanceof CreateApiKeyRequest) {
+                return true;
+            } else if (request instanceof GetApiKeyRequest) {
+                final GetApiKeyRequest getApiKeyRequest = (GetApiKeyRequest) request;
+                if (this.realms.contains("_self") && this.users.contains("_self")) {
+                    return checkIfUserIsOwnerOfApiKeys(authentication, getApiKeyRequest.getApiKeyId(), getApiKeyRequest.getUserName(),
+                            getApiKeyRequest.getRealmName());
+                } else {
+                    return checkIfAccessAllowed(realms, getApiKeyRequest.getRealmName(), realmsPredicate)
+                            && checkIfAccessAllowed(users, getApiKeyRequest.getUserName(), usersPredicate);
+                }
+            } else if (request instanceof InvalidateApiKeyRequest) {
+                final InvalidateApiKeyRequest invalidateApiKeyRequest = (InvalidateApiKeyRequest) request;
+                if (this.realms.contains("_self") && this.users.contains("_self")) {
+                    return checkIfUserIsOwnerOfApiKeys(authentication, invalidateApiKeyRequest.getId(),
+                            invalidateApiKeyRequest.getUserName(), invalidateApiKeyRequest.getRealmName());
+                } else {
+                    return checkIfAccessAllowed(realms, invalidateApiKeyRequest.getRealmName(), realmsPredicate)
+                            && checkIfAccessAllowed(users, invalidateApiKeyRequest.getUserName(), usersPredicate);
+                }
+            }
+            return false;
+        };
+    }
+
+    private boolean checkIfUserIsOwnerOfApiKeys(Authentication authentication, String apiKeyId, String username, String realmName) {
+        if (authentication.getAuthenticatedBy().getType().equals("_es_api_key")) {
+            // API key id from authentication must match the id from request
+            String authenticatedApiKeyId = (String) authentication.getMetadata().get("_security_api_key_id");
+            if (Strings.hasText(apiKeyId)) {
+                return apiKeyId.equals(authenticatedApiKeyId);
+            }
+        } else {
+            String authenticatedUserPrincipal = authentication.getUser().principal();
+            String authenticatedUserRealm = authentication.getAuthenticatedBy().getName();
+            if (Strings.hasText(username) && Strings.hasText(realmName)) {
+                return username.equals(authenticatedUserPrincipal) && realmName.equals(authenticatedUserRealm);
+            }
+        }
+        return false;
```

Comment:
I don't love `/my`, but I also don't have a better suggestion, so lets leave it to a follow up - but I would like you to think about what we can do to make the sorts of errors described above easier to understand.

Reply:
I agree we can leave it for a folllow-up and I uphold my proposal from https://github.com/elastic/elasticsearch/pull/41488 :

> ... we have to decide a "default" value for the user query parameter, depending on the principal's privileges (if it has manage_api_keys the user defaults to all, otherwise it defaults to self). For this we can call has_privileges on the transport handler. That's a fine enough solution, IMO. ...

Get and Invalidate APIs have `id`, `name`, `user` and `realm` REST request parameters. Instead of adding `/my` endpoints to infer the "self" `user` and `realm` I propose to call the `has_privilege` in the REST handler, and to always fill in the `user` and `realm` (if they are not already) so that authz of the transport action always has values to validate.
So, if the `has_privilege` call from the `RestInvalidateApiKeyAction` validates `manage_api_key` then `user` and `realm` are `*` unless otherwise specified. If `has_privilege` validates `manage_own_api_key` then `user` and `realm` are "self" . The authz engine validates the values from the transport request, and the `APIKeyService` translates the values in the query.

Reply:
> I am not very comfortable with the option of interpreting values for user and realm based on the privilege it has as we do not know the intent of the user invoking the API. For eg. when the user does not specify user and realm, and the user has manage_own_api_key we interpret it as self and continue with the operation.

The API is designed to query by `id`, `name`, `user` and `realm` and to have visibility over the whole pool of keys.
If the API can query by `user` and we want to limit it, then we either forbid the parameter in a new API or authorize the parameter value. If we authorize the parameter value, what happens if the parameter is not specified; we either reject it or fill in a default value that will not reject (un-authorize) the request.

I honestly don't know what to do in this case, I would suggest we discuss this as a team.